infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServerCache.java

```diff
@@ -55,12 +57,30 @@ public LogUnitServerCache(LogUnitServerConfig config, StreamLog streamLog) {
 
         MeterRegistryProvider
                 .getInstance()
-                .ifPresent(registry -> CaffeineCacheMetrics.monitor(registry, dataCache, "logunit.read_cache"));
-        MicroMeterUtils.gauge(hitRatioName, dataCache, cache -> cache.stats().hitRate());
-        MicroMeterUtils.gauge(loadTimeName, dataCache, cache -> cache.stats().totalLoadTime());
-        MicroMeterUtils.gauge(weightName, dataCache, cache -> cache.stats().evictionWeight());
+                .ifPresent(registry -> {
+                    CaffeineCacheMetrics.monitor(registry, dataCache, "logunit.read_cache");
+
+                    //calc hit rate
+                    MicroMeterUtils.gauge(hitRatioName, dataCache, cache -> {
+                        double miss = getReadCacheCounter(registry, Tag.of("result", "miss")).count();
+                        double hit = getReadCacheCounter(registry, Tag.of("result", "hit")).count();
+
+                        return hit / (hit + miss);
```

Comment:
Just wondering - If both `miss` and `hit` are 0, it looks like this result will be `NaN`. Is this handled?

Reply:
I would suggest not calculating these manually, but instead use the `CacheStats` object which implements helper methods that account for these edge cases. For example, if  `oldVal = cache.states()` and `newVal = cache.states()`, then you can do `newVal.minutes(oldVal).hitRate()`

Reply:
> Just wondering - If both `miss` and `hit` are 0, it looks like this result will be `NaN`. Is this handled?

I have fixed that issue, thank you for the hint

Reply:
> I would suggest not calculating these manually, but instead use the `CacheStats` object which implements helper methods that account for these edge cases. For example, if `oldVal = cache.states()` and `newVal = cache.states()`, then you can do `newVal.minus(oldVal).hitRate()`
> 
> Note that `cache.states()` takes a snapshot of the stats at that particular time.

In general I agree, but:
 - I don't like the idea of keeping stats snapshot (and it looks like the link to it needs to be mutable)
 - calculating the diff between parsentages between two shapshots seems tedious
 - in my solution we calculate the value out of the numbers we get right away. The calculation itself it pretty simple and straighforward
 - Using my approach we can calculate other numbers also, if needed

Reply:
I don't understand the points. Can you rephrase? (preferably in a more succinct and objective way)

Reply:
It's easy and simple just to take two numbers and divide one by another rather than doing something else

Reply:
Hey Maithem, 
I'he got some logs [here](https://gist.github.com/xnull/2808aa48c27b66e6ccec87cfeeee7af9)
As you can see, the `miss` metric has been aggregated by minute - the value going up and down each minute, which means it really has been aggregated by minute.
The same true for `hit` metric, just by definition.

So, the only thing that I need to do is to divide one number by another (miss vs hit) the same way how it's done in the stats class and I will get hit rate aggregated by minute too.

Reply:
To get non cumulative states there are two ways. Either have a counter that gets reset per-poll period, or you subtract two cumulative counters to get the difference. Instead of showing logs can you write a test case that demonstrates correct per-poll period stats?

Reply:
```java

Optional<FunctionCounter> miss() {
        return MeterRegistryProvider.getInstance()
                .map(registry -> registry.get("cache.gets").tags(missTags).functionCounter());
    }
```

This code snippet will generate a counter with a weak-reference, so if the object is getting gc'd then the counter will reset, but it wouldnt be reporting correct statistics.

Reply:
> ldnt be reporting correct statistics.

I can do that, but that's more of verifying of how Micrometer works underlying. We know that it works in a correct way, otherwise any metric will be affected and the entire thing wouldn't work as expected.

Anyway, let me write a test for that, even though it's a bit complex and time consuming.

Reply:
> ```java
> Optional<FunctionCounter> miss() {
>         return MeterRegistryProvider.getInstance()
>                 .map(registry -> registry.get("cache.gets").tags(missTags).functionCounter());
>     }
> ```
> 
> This code snippet will generate a counter with a weak-reference, so if the object is getting gc'd then the counter will reset, but it wouldnt be reporting correct statistics.

I'm not sure if that gonna be a problem at all. In our hitRatio metrics code we get a  miss and hit counter and use it immediately, I don't see how there could a problem. Even if your scenario happens once in a very rare probability time, we just get a 1.0 at that minute which is not a big of a deal. 
And the same will happen with your approach too, anyway

Reply:
> And the same will happen with your approach too, anyway

False. I think they gauges should have strong references so that issue wouldn't happen.

Reply:
@xnull In a test, it doesn't seem like the logunit_cache_hit_ratio metric changes to 1 in a minute with no activity. Rather, it keeps the same value that it had in previous minutes.

Reply:
From three consecutive minutes (the later two have no activity):

logunit_cache_hit_ratio,id=9812fb56-ec11-4377-a60a-46ebc8ba8a25,metric_type=gauge value=0.607843
logunit_cache_hit_ratio,id=9812fb56-ec11-4377-a60a-46ebc8ba8a25,metric_type=gauge value=0.607843
logunit_cache_hit_ratio,id=9812fb56-ec11-4377-a60a-46ebc8ba8a25,metric_type=gauge value=0.607843